core/src/main/java/org/apache/calcite/rel/metadata/RelMdForeignKeys.java

```diff
@@ -0,0 +1,246 @@
+    for (int i = 0; i < projExprs.size(); i++) {
+      RexNode projExpr = projExprs.get(i);
+      if (projExpr instanceof RexInputRef) {
+        int inputIndex = ((RexInputRef) projExpr).getIndex();
+        inToOutIndexBuilder.put(inputIndex, i);
+        inColumnsBuilder.set(inputIndex);
+      }
+    }
+    final ImmutableBitSet inColumnsUsed = inColumnsBuilder.build();
+    if (inColumnsUsed.isEmpty()) {
+      return EMPTY_BIT_SET;
+    }
+
+    final Map<Integer, ImmutableBitSet> mapInToOutPos =
+        Maps.transformValues(inToOutIndexBuilder.build().asMap(), ImmutableBitSet::of);
+    final ImmutableBitSet inputForeignKeys =
+        mq.getForeignKeys(rel.getInput(), ignoreNulls);
+    if (inputForeignKeys.isEmpty()) {
+      return EMPTY_BIT_SET;
+    }
+
+    ImmutableBitSet allOutColumns = ImmutableBitSet.of();
+    for (int index : inputForeignKeys.asList()) {
+      ImmutableBitSet outColumns = mapInToOutPos.get(index);
+      if (outColumns != null && !outColumns.isEmpty()) {
+        allOutColumns = allOutColumns.union(outColumns);
+      }
+    }
+    return allOutColumns.isEmpty() ? EMPTY_BIT_SET : allOutColumns;
+  }
+
+  public ImmutableBitSet getForeignKeys(TableScan rel, RelMetadataQuery mq,
+      boolean ignoreNulls) {
+    final RelOptTable table = rel.getTable();
+    final BuiltInMetadata.ForeignKeys.Handler handler =
+        table.unwrap(BuiltInMetadata.ForeignKeys.Handler.class);
+    if (handler != null) {
+      return handler.getForeignKeys(rel, mq, ignoreNulls);
+    }
+
+    final List<RelReferentialConstraint> referentialConstraints =
+        table.getReferentialConstraints();
+    if (referentialConstraints == null || referentialConstraints.size() == 0) {
+      return EMPTY_BIT_SET;
+    }
+    final List<IntPair> foreignUniquePair = referentialConstraints.stream()
+        .map(RelReferentialConstraint::getColumnPairs)
+        .flatMap(Collection::stream)
+        .collect(Collectors.toList());
+
+    List<Integer> foreignKeys = IntPair.left(foreignUniquePair);
+    if (!ignoreNulls) {
+      final List<RelDataTypeField> fieldList = rel.getRowType().getFieldList();
+      foreignKeys = foreignKeys.stream()
+          .filter(index -> !fieldList.get(index).getType().isNullable())
+          .collect(Collectors.toList());
+    }
+    return foreignKeys.isEmpty() ? EMPTY_BIT_SET : ImmutableBitSet.of(foreignKeys);
+  }
+
+  public ImmutableBitSet getForeignKeys(SetOp rel, RelMetadataQuery mq,
+      boolean ignoreNulls) {
+
+    ImmutableBitSet foreignKeys = ImmutableBitSet.of();
+    for (RelNode input : rel.getInputs()) {
+      ImmutableBitSet inputForeignKeys = mq.getForeignKeys(input, ignoreNulls);
+      if (inputForeignKeys.isEmpty()) {
+        return EMPTY_BIT_SET;
+      }
+      foreignKeys = foreignKeys.isEmpty()
+          ? inputForeignKeys : foreignKeys.intersect(inputForeignKeys);
+    }
+    if (foreignKeys.isEmpty()) {
```

Comment:
I think this is redundant, if `isEmpty()` returns true, no need to return an empty bitset

Reply:
Yeah, it is also redundant somewhere else. I will fix them